File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        n = len(arr)
        if n == 0: return []

        # clone the array and sort it
        sorted_arr = sorted(arr)

        # assign rank to each number
        rank = {}
        rank[sorted_arr[0]] = 1
        curr = 2
        for i in range(1, n):
            if sorted_arr[i] == sorted_arr[i-1]: continue
            rank[sorted_arr[i]] = curr
            curr += 1

        # build the final output
        for i, num in enumerate(arr):
            arr[i] = rank[num]
        
        return arr

File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.cs
public class Solution {
    public int[] ArrayRankTransform(int[] arr) {
        var n = arr.Length;
        if(n == 0) return new int[0];

        // clone the array and sort it
        var sorted = (int[])arr.Clone();
        Array.Sort(sorted);

        // assign rank to each number
        var rank = new Dictionary<int, int>();
        rank[sorted[0]] = 1;
        for(int i=1, curr=1; i<n; i++){
            if(sorted[i] == sorted[i-1]) continue;
            rank[sorted[i]] = ++curr;
        }

        // build the final output
        var ans = new int[n];
        for(var i=0; i<n; i++)
            ans[i] = rank[arr[i]];
        
        return ans;
    }
}